File: kotlin/src/test/kotlin/com/imogen/sdk/AssetsTest.kt
package com.imogen.sdk

import kotlinx.coroutines.test.runTest
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * The id-list overload of the bulk asset mutations is the original, shorter way of
 * saying "these assets" -- it must keep producing a bare `{"assetIds":[...]}` body, not
 * grow the `query`/`except` fields a selection carries. All six call sites (trash,
 * restore, addAssets, removeAssets, moveIn, moveOut) delegate through the same
 * `AssetSelection(assetIds = ...)` path, so one covers the shape for all of them.
 */
class AssetsTest {

    @Test
    fun `trash accepts a bare id list, as it always has`() = runTest {
        val stub = Stub { _, _ -> Reply.json("""{"count":1}""") }

        ImogenClient(
            ClientOptions(baseUrl = "https://photos.example.test", engine = stub.engine)
        ).use { imogen ->
            imogen.assets.trash(listOf("6a5f1e2c-90b4-4d1a-8f3e-2b7c9d0a1e45"))
        }

        assertEquals(
            """{"assetIds":["6a5f1e2c-90b4-4d1a-8f3e-2b7c9d0a1e45"]}""",
            String(stub.calls.single().body),
        )
    }
}

File: typescript/packages/sdk/src/assets.test.ts
import { expect, test } from 'bun:test'
import { Assets } from './assets.ts'
import type { HttpClient, RequestOptions } from './http.ts'

type RecordedCall = {
  method: string
  path: string
  query?: Record<string, unknown>
  body?: unknown
}

function recordingHttp(payload: unknown): HttpClient & { calls: RecordedCall[] } {
  const calls: RecordedCall[] = []
  const http = {
    calls,
    baseUrl: 'https://example.test',
    url: (path: string) => path,
    request: async (method: string, path: string, options: RequestOptions = {}) => {
      calls.push({ method, path, query: options.query, body: options.body })
      return payload
    },
    send: async () => {
      throw new Error('recordingHttp: send() not implemented')
    },
  }
  return http as unknown as HttpClient & { calls: RecordedCall[] }
}

test('timelineBucket asks for one period', async () => {
  const http = recordingHttp({ items: [], nextCursor: null, total: 0 })
  await new Assets(http).timelineBucket({ period: '2011-08', limit: 5000 })
  expect(http.calls[0]).toMatchObject({
    method: 'GET',
    path: '/api/v1/assets/timeline/bucket',
    query: { period: '2011-08', limit: 5000 },
  })
})

test('trash accepts a bare id list, as it always has', async () => {
  const http = recordingHttp({ count: 1 })
  await new Assets(http).trash(['6a5f1e2c-90b4-4d1a-8f3e-2b7c9d0a1e45'])
  expect(http.calls[0]?.body).toEqual({ assetIds: ['6a5f1e2c-90b4-4d1a-8f3e-2b7c9d0a1e45'] })
})

test('trash accepts a selection by query', async () => {
  const http = recordingHttp({ count: 12431 })
  await new Assets(http).trash({ query: { favorite: true }, except: [] })
  expect(http.calls[0]?.body).toEqual({ query: { favorite: true }, except: [] })
})
